File: problems/C++/116. Populating Next Right Pointers in Each Node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return nullptr;
        
        Node* l = root;
        while (l->left) {
            Node* h = l;
            while (h) {
                h->left->next = h->right;
                if (h->next) {
                    h->right->next = h->next->left;
                }
                h = h->next;
            }
            l = l->left;
        }
        return root;
    }
};

File: problems/Java/116. Populating Next Right Pointers in Each Node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) return null;
        
        Node l = root;
        while (l.left != null) {
            Node h = l;
            while (h != null) {
                h.left.next = h.right;
                if (h.next != null) {
                    h.right.next = h.next.left;
                }
                h = h.next;
            }
            l = l.left;
        }
        return root;
    }
}